layerfetcher/layers.go

```diff
@@ -403,74 +264,80 @@ func (l *Logic) PollLayerBlocks(ctx context.Context, layerID types.LayerID, hash
 		// TODO: fail sync?
 		l.log.WithContext(ctx).With().Debug("error fetching layer blocks", layerID, log.Err(err))
 	}
 
-	if len(blocks.VerifyingVector) > 0 {
+	if len(blocks.InputVector) > 0 {
 		l.log.WithContext(ctx).With().Debug("saving input vector from peer blocks", layerID)
-		err := l.layerDB.SaveLayerInputVectorByID(ctx, layerID, blocks.VerifyingVector)
+		err := l.layerDB.SaveLayerInputVectorByID(ctx, layerID, blocks.InputVector)
 		if err == nil {
 			return
 		}
 		l.log.WithContext(ctx).With().Error("failed to save input vector", layerID, log.Err(err))
-	}
-
-	if !layerID.GetEpoch().IsGenesis() {
+	} else if !layerID.GetEpoch().IsGenesis() {
 		if err := l.GetInputVector(ctx, layerID); err != nil {
 			l.log.WithContext(ctx).With().Debug("error getting input vector", layerID, log.Err(err))
 		}
 	}
 }
 
-// receiveBlockHashes is called when response of block IDs for a layer hash is received from remote peer.
+func extractPeerResult(logger log.Log, layerID types.LayerID, data []byte, peerErr error) (result peerResult) {
+	result.err = peerErr
+	if peerErr != nil {
+		logger.With().Debug("received peer error for layer content", layerID, log.Err(peerErr))
+		return
+	}
+
+	var blocks layerBlocks
+	if err := types.BytesToInterface(data, &blocks); err != nil {
+		logger.With().Debug("error converting bytes to layerBlocks", log.Err(err))
+		result.err = err
+		return
+	}
+
+	result.data = &blocks
+	if len(blocks.Blocks) == 0 {
+		result.err = ErrZeroLayer
+	}
+	// TODO check layer hash to be consistent with the content. if not, blacklist the peer
+	return
+}
+
+// receiveLayerContent is called when response of block IDs for a layer hash is received from remote peer.
 // if enough responses are received, it notifies the channels waiting for the layer blocks result.
-func (l *Logic) receiveBlockHashes(ctx context.Context, layerID types.LayerID, peer peers.Peer, expectedResults int, data []byte, extErr error) {
-	pRes := &peerResult{err: extErr}
-	if extErr != nil {
-		if !errors.Is(extErr, ErrZeroLayer) || !(layerID.GetEpoch()).IsGenesis() {
-			l.log.WithContext(ctx).With().Error("received error from peer for block hashes", log.Err(extErr), layerID, layerID.GetEpoch())
-		}
-	} else if data != nil {
-		var blocks layerBlocks
-		convertErr := types.BytesToInterface(data, &blocks)
-		if convertErr != nil {
-			l.log.WithContext(ctx).With().Error("error converting bytes to layerBlocks", log.Err(convertErr))
-			pRes.err = convertErr
-		} else {
-			pRes.data = data
-			l.fetchLayerBlocks(ctx, layerID, &blocks)
-		}
+func (l *Logic) receiveLayerContent(ctx context.Context, layerID types.LayerID, peer peers.Peer, expectedResults int, data []byte, peerErr error) {
+	l.log.WithContext(ctx).With().Debug("received layer content from peer", log.String("peer", peer.String()))
+	peerRes := extractPeerResult(l.log.WithContext(ctx), layerID, data, peerErr)
+	if peerRes.err == nil {
+		l.fetchLayerBlocks(ctx, layerID, peerRes.data)
```

Comment:
will this try to fetch same blocks if they were received from different peers?

Reply:
no. this happens in fetcher.go, which groups by hash. if there is already a request for the same hash pending, it will wait for the pending request for the same hash.